Updates to Stub Reconstruction

gputnam opened the pull request:
A few various updates to stub reconstruction. Doesn't affect anything currently run in production.

Depends on https://github.com/SBNSoftware/sbnobj/pull/13.

wesketchum commented on sbncode/TPCReco/VertexStub/PlaneTransform.h:
```diff
@@ -0,0 +1,89 @@
+class PlaneTransform {
+public:
+  PlaneTransform(fhicl::ParameterSet const& p);
+  double UVtoW(const double u, const double v) const;
+  double VWtoU(const double v, const double w) const;
+  double WUtoV(const double w, const double u) const;
+  
+  double UVtoY(const double u, const double v) const;
+  double UVtoZ(const double u, const double v) const;
+  double UWtoY(const double u, const double w) const;
+  double UWtoZ(const double u, const double w) const;
+  double VWtoY(const double v, const double w) const;
+  double VWtoZ(const double v, const double w) const;
+  
+  double YZtoU(const double y, const double z) const;
+  double YZtoV(const double y, const double z) const;
+  double YZtoW(const double y, const double z) const;
```
@PetrilloAtWork may know better, but I think there should be Geometry utilities that can do these, rather than making new functions and relying on a properly configured fcl config. Is it possible to rely on the geometry service here?

gputnam replied:
I looked and could not find any Geometry functions that provided this utility. Do you know where it would be?

wesketchum replied:
Yeah, @PetrilloAtWork should definitely weigh in ... but I think at least grabbing wire angle in a plane is here:

https://github.com/SBNSoftware/larcorealg/blob/develop/larcorealg/Geometry/PlaneGeo.h#L513

and PlaneGeo has a lot of other utilities to use I think ... I'm not sure this is exactly these functions, but there's a lot to be able to pull from the geometry itself rather than rely on fcl config.

Even while leaving these functions/etc., I think at a minimum it'd be good to not have specified in fcl the different angles and views, as that should be coming from the geometry itself.

gputnam replied:
I also have a little trouble wrapping my head around this stuff, but I don't believe that this code makes that assumption. 

This is because the "Wire-Coordinate" that is passed to these functions is not the same as the wire number. The wire coordinate of a wire is obtained from the sbn::PlaneTransform::WireCoordinate function, which explicitly pulls from the wire geometry. This means that the origin used in the plane matching will be the same as the origin in the detector geometry -- it is not assumed to be in any special location.

wesketchum commented on sbncode/TPCReco/VertexStub/StubBuilder.cxx:
```diff
@@ -0,0 +1,238 @@
+  float hit_x = dprop.ConvertTicksToX(vhit_hit.PeakTime(), vhit_hit.WireID());
+  geo::PlaneID vhit_plane = vhit_hit.WireID();
+
+  std::cout << "COLLECTING HITS!\n";
+  std::cout << "HIT W: " << hit_w << " X: " << hit_x << std::endl;
+  std::cout << "VTX W: " << vert_wf << " X: " << vert_x << std::endl;
```
Here (and other places): should remove couts or put them behind some verbosity flag in the config.

gputnam replied:
Done.

wesketchum commented on sbncode/TPCReco/VertexStub/TwoPlaneStubMerge_tool.cc:
```diff
@@ -0,0 +1,220 @@
+  if (A_nhit != B_nhit) return A_nhit > B_nhit;
+
+  // Sort by desirability of plane: 2 better than 0 better than 1
+  return ((A.vhit_hit->WireID().Plane + 1) % 3) < ((B.vhit_hit->WireID().Plane + 1) % 3);
```
Do we want to rely on the numbering scheme here, or to check for 'collection vs not? or to make it configurable what plans to set as desirable? I'm just not sure if 0 vs. 1 holds up over time.

gputnam replied:
The 1st induction plane is also better than the second. Is there a way to express that without relying on the Plane number?

gputnam replied:
Two other notes: this piece of code isn't making any analysis decision -- it just picks which stub goes first in the output vector. So it does not matter if the ordering doesn't hold up over time.

Also, I don't see any reason to expect that the 2nd induction would perform better than the first in ICARUS, SBND, or uBooNE.

wesketchum replied:
Well, but I thought that the output vector was ordered such that the "best" plane was the first in the list? Or I am likely misunderstanding something...maybe that was a different output vector in the Stub?

But, that's fair enough. It would be good if this was documented. Is there a presentation or something that we can reference for the future?

gputnam replied:
That's correct, it tries to place the best plane first. But it doesn't need to do that perfectly.

This is the most recent presentation on what this code does: https://sbn-docdb.fnal.gov/cgi-bin/private/ShowDocument?docid=21315.

PetrilloAtWork commented on sbncode/TPCReco/NuVertexChargeTree_module.cc:
```diff
@@ -504,33 +648,57 @@ void sbn::NuVertexChargeTree::FillStubPFPs(
+    const std::vector<sbn::StubInfo> &stubs,
     const std::vector<unsigned> &stub_hit_inds,
-    const std::vector<std::vector<art::Ptr<recob::Hit>>> &stub_hits,
+    const geo::GeometryCore *geo,
```
For uniformity with the other services, I suggest `geo::GeometryCore const& geom` (i.e. a reference rather than a pointer): I am assuming here that the code will not work if `geom` is null.
Also, I suggest renaming `geo` to `geom`, since `geo` is already a namespace: the compiler seems to understand, but it's a bit confusing.

gputnam replied:
Is there a problem with using a variable name that happens to line up with a namespace? As you said, it's acceptable for C++.

Also, the service function getter returns a pointer, so it's natural to use a pointer to carry "geo/geom" around. I don't think switching from a pointer would change any behavior of the code.

PetrilloAtWork commented on sbncode/TPCReco/VertexStub/VertexStubTracker_module.cc:
```diff
@@ -181,95 +120,71 @@ void sbn::VertexStubTracker::produce(art::Event& e)
-      charge += fCaloAlg.ElectronsFromADCArea(hit.Integral(), hit.WireID().Plane) * fCaloAlg.LifetimeCorrection(clock_data, dprop, hit.PeakTime(), 0.);
-      wires.insert(hit.WireID().Wire);
+    // Run all of the merging tools
+    for (unsigned i_mrg = 0; i_mrg < fStubMergeTools.size(); i_mrg++) {
+      stubs = fStubMergeTools[i_mrg]->Merge(stubs, geo, sce, clock_data, dprop);
```
Usual suggestion to use range-for loop to avoid index mistakes:
```suggestion
    for (auto const& mergeTool: fStubMergeTools) {
      stubs = mergeTool->Merge(stubs, geo, sce, clock_data, dprop);
```

gputnam replied:
I think it's fine as is -- there are no mistakes made with the index here.

PetrilloAtWork commented on sbncode/TPCReco/VertexStub/VertexStubTracker_module.cc:
```diff
@@ -181,95 +120,71 @@ void sbn::VertexStubTracker::produce(art::Event& e)
-    }
-
-  } // end iterate over vertex hits
+    for (unsigned i_stub = 0; i_stub < stubs.size(); i_stub++) {
+      const sbn::StubInfo &sinfo = stubs[i_stub];
```
Yeah, I really like range-for loops:
```suggestion
    for (const sbn::StubInfo &sinfo: stubs) {
```

PetrilloAtWork commented on sbncode/TPCReco/VertexStub/VertexStubTracker_module.cc:
```diff
@@ -181,95 +120,71 @@ void sbn::VertexStubTracker::produce(art::Event& e)
+      outStubs->push_back(sinfo.stub);
+      art::Ptr<sbn::Stub> outStub = StubPtrMaker(outStubs->size() - 1);
+      assn->addSingle(sinfo.vhit, outStub);
+      for (unsigned i_hit = 0; i_hit < sinfo.hits.size(); i_hit++) { 
+        hitAssn->addSingle(outStub, sinfo.hits[i_hit]);
```
I really _really_ do.
```suggestion
      for (auto const& hit: sinfo.hits) { 
        hitAssn->addSingle(outStub, hit);
```

PetrilloAtWork reviewed (changes requested):
So here I am with 70 comments... probably the next time we meet in person you'll punch my face.
Most of my comments are about coding style and choices.
I do have some _recommendations_, including documenting methods and data members in Doxygen (yes, I know how painful it is) and avoiding `std::cout`, some of them stronger, like using `art::Event` access that throws exception and using `geo::GeometryCore` service provider when possible.

But I think there are a few show-stoppers for code that smells like not being compatible with ICARUS geometry. Using `geo::GeometryCore` methods greatly helps in there, but there are also possibly hidden assumptions that need investigation.
In terms of "geometry code from Pandora", my vague recollection is that ICARUS had to wait for Pandora geometry code to be adapted to horizontal wires before properly working: this is to say that it is not impossible that its geometry code is not compatible.

Needless to say, I am going to try to help with these changes.

gputnam commented on sbncode/TPCReco/NuVertexChargeTree_module.cc:
```diff
@@ -706,11 +936,74 @@ void sbn::NuVertexChargeTree::FillNeutrino(const simb::MCTruth &nu,
     fFSPVisE.push_back(thisVisE / 3. /* average over each plane */);
+    fFSPQ.push_back(thisQ / 3.);
+
+    for (unsigned v = 0; v < 3; v++) {
```
Yeah I was a bit sloppy here writing stuff for SBND.

There are two future tasks I have for myself -- get recent developments working (1) in ICARUS and (2) with Space Charge. I'd consider that work outside the scope of this pull request though.

I'd ask that I still be able to merge stuff in despite with these issues in place. My immediate plans are to use this stuff in ICARUS, so it's something I will plan to do soon.

PetrilloAtWork replied:
Can you collect the ICARUS fixes in a to-do list, so that they don't get lost?

gputnam replied:
Done!

gputnam commented on sbncode/TPCReco/NuVertexChargeTree_module.cc:
```diff
@@ -706,11 +936,74 @@ void sbn::NuVertexChargeTree::FillNeutrino(const simb::MCTruth &nu,
+    geo::Vector_t start_dir(particle.Momentum().Vect().Unit());
+    geo::TPCID firstTPC = geo->FindTPCAtPosition(start_loc);
+    if (firstTPC) {
+      fFSPPitchU.push_back(sbn::GetPitch(geo, sce, start_loc, start_dir, (geo::View_t)0, firstTPC, true, true));
+      fFSPPitchV.push_back(sbn::GetPitch(geo, sce, start_loc, start_dir, (geo::View_t)1, firstTPC, true, true));
+      fFSPPitchW.push_back(sbn::GetPitch(geo, sce, start_loc, start_dir, (geo::View_t)2, firstTPC, true, true));
```
Same as above comment.

PetrilloAtWork replied:
Same comment as above: mark this in the list, then resolve this comment, please.

gputnam commented on sbncode/TPCReco/NuVertexChargeTree_module.cc:
```diff
@@ -850,8 +1145,8 @@ void sbn::NuVertexChargeTree::analyze(art::Event const& evt)
         const std::vector<art::Ptr<sbn::Stub>> &this_vh = vertexStubs.at(i_stub);
-        if (this_vh.size()) {
-          sliceStubs.back().push_back(this_vh.at(0));
+        for (unsigned i_vh_stub = 0; i_vh_stub < this_vh.size(); i_vh_stub++) {
+          sliceStubs.back().push_back(this_vh.at(i_vh_stub));
```
Although the conventional for loop is not strictly required here, I prefer it and think it is easier to conceptualize and consistent with the surrounding code.

The check is redundant but it's not a piece of the code that needs to be very performant.

PetrilloAtWork replied:
The easier path to conceptualisation is subjective, so I honour yours.
One note about performance: I agree we are not hunting performance here. In general, however, keep in mind that `at()` pulls in additional code to allow for exception handling, which may be a huge absolute difference. My point here is that the additional impact goes far beyond a `if (i < size())` check. This said, in most places you will not notice any performance difference anyway.

gputnam commented on sbncode/TPCReco/VertexStub/CMakeLists.txt:
```diff
@@ -1,3 +1,30 @@
+art_make_library(LIBRARY_NAME sbncode_StubHelpers
```
I'm not sure what the alternative is here. It's a library that contains a few .cxx files, I think if the name is any one in particular then it is confusing.

I think it's good that the library name is explicit. This way you can look at the CMake file and see what the library is, instead of relying on a convention that is only sometimes followed in larsoft.

PetrilloAtWork replied:
This was an attempt to make that convention more followed. LArSoft almost universally follows it (exceptions are a handful). ICARUS as well (that probably just proves that I really think it's a good practice...). SBND somehow less.

The alternative is conventional, to use the name of the directory. The compromise to pick a name is to put the few `.cxx` files in their own subdirectory (e.g. `sbncode/TPCReco/VertexStub/Helpers`).

Strangely enough, the meaningfulness of the library name has not been that paramount in my experience: that name only enters my life when I have written the source code and I need the names of the libraries connected to the headers I included. The most automatic the conversion header → library name is, the better for me. Maybe your experience is different?
One more thing before I stop rambling: again from my experience, following a convention makes life of maintainers much, much easier.

About "can look at CMake file": CMake files are not distributed with the code, and to look at it one needs to check out the repository or use web browsing (admittedly, GitHub makes the latter easy enough).

gputnam replied:
Ok, that makes sense. I will change it to use the conventional library name.

gputnam commented on sbncode/TPCReco/VertexStub/PlaneTransform.cxx:
```diff
@@ -0,0 +1,168 @@
+  if (abs(m_sinVminusU) < std::numeric_limits<double>::epsilon() ||
+      abs(m_sinWminusV) < std::numeric_limits<double>::epsilon() ||
+      abs(m_sinUminusW) < std::numeric_limits<double>::epsilon()) {
+    std::cout << "PlaneTransform::PlaneTransform -- does not support provided TPC configuration.\n";
```
I copied the way that it is currently handled in Pandora.

gputnam replied:
Changed to an exception --fixed.

gputnam commented on sbncode/TPCReco/VertexStub/StubBuilder.cxx:
```diff
@@ -0,0 +1,232 @@
+  }
+
+  for (unsigned i_pfp = 0; i_pfp < pfps.size(); i_pfp++) {
+    if (pfp_nhit[i_pfp] >= hits.size() / 2) return i_pfp;
```
If more than half of the hits in the stub match to a single PFParticle, the Stub "matches" to that particular PFParticle.

PetrilloAtWork replied:
Consider adding that as a comment.
Well, now it seems clear to me :-D
Resolved on my side.

gputnam replied:
Added a comment.

gputnam commented on sbncode/TPCReco/VertexStub/StubBuilder.cxx:
```diff
@@ -0,0 +1,232 @@
+  geo::PlaneID vhit_plane = vhit_hit.WireID();
+
+  // get the line-segment slope / intercept
+  float slope = (hit_x - vert_x) / (hit_w - vert_wf);
```
I don't know if that's a good idea. 

I think that it may happen sometimes that the hit wire is the same as the vertex wire, and the code should not crash in that case.

PetrilloAtWork replied:
What should it do in that case?

gputnam replied:
Ok -- then only hits on the (shared) vertex + hit-end-point wire should be included.

I have added in logic to do this.

gputnam commented on sbncode/TPCReco/VertexStub/StubBuilder.cxx:
```diff
@@ -0,0 +1,232 @@
+  fSliceTrks.clear();
+
+  // Build the map of the slice key to the hits
+  art::Handle<std::vector<recob::Slice>> slice_handle;
+  e.getByLabel(pfplabel, slice_handle);
```
I'd say that's a feature, not a bug. I don't know if it's always a good idea to crash if slices are missing. Just not producing any output (which is what this code does) I think is reasonable.

cjbackhouse replied:
Just a drive-by to drop in my 2c here.
I don't think "just don't produce any output if you can't find the slices" is a good idea. If the user meant to not specify valid slices, they could use a blank label (and it would be appropriate to have special handling for that).

Scientific software isn't an autopilot or a pacemaker, we have no obligation to make a best effort when there are problems. The much bigger risk is that we do months of production processing and then discover no stubs were made because of a trivial misconfiguration, which the code covered up.

Similarly for things that "should never happen". The right thing is to blow up the job with a verbose explanation of the problem so it can be understood and fixed properly, not sweep it under the rug. Not saying you're guilty of this, but I've definitely seen a lot of it in physics code and it drives me mad.

PetrilloAtWork replied:
> not [...] a pacemaker, we have no obligation to make a best effort when there are problems

I will definitely feel safe if you ever decide to go to medical.

gputnam commented on sbncode/TPCReco/VertexStub/StubBuilder.cxx:
```diff
@@ -0,0 +1,232 @@
+
+  for (unsigned i_slc = 0; i_slc < slices.size(); i_slc++) {
+    const std::vector<art::Ptr<recob::PFParticle>> thisSlicePFPs = slicePFPs.at(i_slc);
+    art::FindManyP<recob::Cluster> pfparticleClusters(thisSlicePFPs, e, pfplabel);
```
Ok! The code so far is pretty quick, but I will keep that in mind if it seems to start to slow down.

gputnam commented on sbncode/TPCReco/VertexStub/StubBuilder.h:
```diff
@@ -0,0 +1,63 @@
+#include "lardataobj/RecoBase/Wire.h"
+#include "lardataobj/RecoBase/Hit.h"
+#include "lardataobj/RecoBase/TrackHitMeta.h"
+#include "larcore/Geometry/Geometry.h"
```
Those objects are used in the code, I don't think these are superfluous.

PetrilloAtWork replied:
`larcore/Geometry/Geometry`?
(note that these comments refer only to the last line, unless the comment box specifies a range)

gputnam replied:
Ah, ok. Well I have removed that line. So I will resolve this.

gputnam commented on sbncode/TPCReco/VertexStub/StubMergeAlgorithms.cxx:
```diff
@@ -0,0 +1,245 @@
+#include "StubMergeAlgorithms.h"
+
+geo::Point_t sbn::GetLocation(const spacecharge::SpaceCharge *sce, geo::Point_t loc_w, unsigned TPC, float xsign) {
+  if (sce && sce->EnableCalSpatialSCE()) {
+    geo::Vector_t offset = sce->GetCalPosOffsets(loc_w, TPC);
+    loc_w.SetX(loc_w.X() + xsign * offset.X());
+    loc_w.SetY(loc_w.Y() + offset.Y());
+    loc_w.SetZ(loc_w.Z() + offset.Z());
+  }
+
+  return loc_w;
+}
```
This is based on code I originally wrote for the Calorimetry.

The problem is that the x-sign convention is not consistent across different implementations/versions of the SpaceCharge service, and it's something we've had issues with in the past. So I include the added flexibility to handle this case.

PetrilloAtWork replied:
Groan. Please if you find again this kind of issue report it... it seems the messy `SpaceCharge` interface is never going to be sorted out :'(

gputnam commented on sbncode/TPCReco/VertexStub/StubMergeAlgorithms.cxx:
```diff
@@ -0,0 +1,245 @@
+    float xx = loc.X();
+    if (xx < 0) { corr = -1; }
+
+    // for some reason, one needs to flip the sign of the x-direction when correcting for field distortion
+    geo::Vector_t offset = sce->GetPosOffsets(loc);
+    loc.SetX(loc.X() + corr * xsign * offset.X());
+    loc.SetY(loc.Y() + offset.Y());
+    loc.SetZ(loc.Z() + offset.Z());
+  }
```
I'm not sure what you mean -- this is that exact helper function.

PetrilloAtWork replied:
I meant: put the six lines into a function; I am under the impression that these six lines are already in an helper function, `sbn::GetLocation()`, but I might have missed some detail.
If `sbn::GetLocation()` is indeed already that, then I suggest you replace these six lines with a call to it.

gputnam replied:
Gotcha -- I have fixed that code to use the helper function.

gputnam commented on sbncode/TPCReco/VertexStub/StubMergeAlgorithms.cxx:
```diff
@@ -0,0 +1,245 @@
+  // seen by the wire planes
+  if (sce && sce->EnableCalSpatialSCE() && correct_sce && track_is_sce_corrected) {
+    // fix negative sign in TPC 0
+    int corr = 1;
+    float xx = loc.X();
+    if (xx < 0) { corr = -1; }
```
Agreed -- I will fix this when I port stuff to Space-Charge/ICARUS.

PetrilloAtWork replied:
Mark it in the list, then resolve this comment, please.

gputnam replied:
Done!

gputnam commented on sbncode/TPCReco/VertexStub/StubMergeAlgorithms.cxx:
```diff
@@ -0,0 +1,245 @@
+    geo::Point_t loc_mdx = loc - dir * (geo->WirePitch(view) / 2.);
+    geo::Point_t loc_pdx = loc + dir * (geo->WirePitch(view) / 2.);
+
+    geo::Vector_t loc_mdx_offset = sce->GetPosOffsets(loc_mdx);
+
+    loc_mdx.SetX(loc_mdx.X() + corr * xsign * loc_mdx_offset.X());
+    loc_mdx.SetY(loc_mdx.Y() + loc_mdx_offset.Y());
+    loc_mdx.SetZ(loc_mdx.Z() + loc_mdx_offset.Z());
+
+    geo::Vector_t loc_pdx_offset = sce->GetPosOffsets(loc_pdx);
+    loc_pdx.SetX(loc_pdx.X() + corr * xsign * loc_pdx_offset.X());
+    loc_pdx.SetY(loc_pdx.Y() + loc_pdx_offset.Y());
+    loc_pdx.SetZ(loc_pdx.Z() + loc_pdx_offset.Z());
```
I think the repetition is fine here. The calculation of pitch in the presence of space charge is complicated, and I think it's good to be spelled out explicitly.

PetrilloAtWork replied:
You can spell it out explicitly in one function, giving that function a nice name (`applySpaceChargeOffset()`?), and call that function twice. Personally I find it clearer than reading the lower level code.

PetrilloAtWork replied:
Anyway, I let you decide how to proceed.
You can resolve this conversation if you like.

gputnam replied:
On second glance I agree with your reasoning -- I have fixed this.

gputnam commented on sbncode/TPCReco/VertexStub/StubMergeAlgorithms.cxx:
```diff
@@ -0,0 +1,245 @@
+  return loc;
+}
+
+double sbn::GetPitch(
```
This is the same as the Pitch calculation that is done in the calorimetry. It needs to handle a few different cases (presence of the space charge or not; whether the absolute positions of the PFParticle are changed).

The key thing though is that the presence of space charge physically changes the pitch beyond the correction of the absolute positions on either end of the wire.

I guess there is a function in the geometry that one can "outsource" some of the math to. I don't think it is necessarily advantageous for me to use it. The math I have written in this function is not wrong, and I can see what it is doing by direct inspection. And since it's just a piece of math, I don't anticipate having to change it in the future.

gputnam commented on sbncode/TPCReco/VertexStub/StubMergeAlgorithms.cxx:
```diff
@@ -0,0 +1,245 @@
+  if (A.vhit_hit->WireID().Plane != B.vhit_hit->WireID().Plane) return false;
+
+  for (const art::Ptr<recob::Hit> &h: A.hits) {
+    if (A.stub.OnCore(h->WireID())) {
+      if (h.key() == B.vhit_hit.key()) {
```
I don't think performance is very important here, and the code is more clear as it is written.

gputnam commented on sbncode/TPCReco/VertexStub/StubMergeAlgorithms.cxx:
```diff
@@ -0,0 +1,245 @@
+  geo::Point_t vertex(A.stub.vtx);
+  
+  // project the vertex onto the wireplane
+  float vert_w = geo->WireCoordinate(vertex, A.vhit_hit->WireID()) * geo->WirePitch();
```
I think the code is more clear as written. There is a vast amount of helper functions in the geometry and multiple different ways to do the same thing. This is the convention I have stuck with and I think it is clear.

"WireCoordinate" makes it clear that we are talking about wire numbers, and then the multiplication by WirePitch() makes it clear that the units are in cm. Changing it to "PlaneCoordinate" obscures both of those facts.

PetrilloAtWork replied:
It also assumes that all pitches are the same. Not a bit deal, until you try to port this code to DUNE.

gputnam commented on sbncode/TPCReco/VertexStub/StubMergeAlgorithms.cxx:
```diff
@@ -0,0 +1,245 @@
+  float vert_w = geo->WireCoordinate(vertex, A.vhit_hit->WireID()) * geo->WirePitch();
+  float vert_x = vertex.x();
+
+  float Ahit_w = A.vhit_hit->WireID().Wire * geo->WirePitch() - vert_w;
```
Again I'm not too worried about either of these things. This not a piece of the code where the performance is particularly important.

And I am very much not concerned if this code can run without issue in DUNE. I'm writing code for SBN -- I think worrying about a detector that is currently unspecified is not a productive use of my time; I have no idea what things may or may not have to change for DUNE, probably a lot of things.

gputnam commented on sbncode/TPCReco/VertexStub/StubMergeAlgorithms.cxx:
```diff
@@ -0,0 +1,245 @@
+  // Correct for space charge
+  pos = GetLocation(sce, pos, A.vhit_hit->WireID().TPC);
+
+  if (VERBOSE) {
```
I don't think it's necessary here. 

It's information that you would only print out to do low level debugging. I think it's fine to hide the printouts behind a switch that you would have to recompile in order to change.

gputnam commented on sbncode/TPCReco/VertexStub/PlaneTransform.cxx:
```diff
@@ -0,0 +1,168 @@
+  return -100000.;
+}
+
+double sbn::PlaneTransform::WireCoordinate(const geo::GeometryCore *geo, const geo::WireID &w) const {
+  geo::View_t v = geo->View(w);
+  TVector3 coord = geo->Wire(w).GetCenter();
+
+  return YZtoPlane(v, coord.Y(), coord.Z());
+}
```
This piece of code is the key to my response to an earlier comment. This code is different from just using the wire number -- it gets the location of the wire along the wireplane direction in detector coordinates. This is how the cross-plane matching code works.

gputnam commented on sbncode/TPCReco/VertexStub/StubMergeAlgorithms.cxx:
```diff
@@ -0,0 +1,245 @@
+  static const bool VERBOSE = false;
+
+  // look up the wire-coordinate of both wires
+  double A_w = T.WireCoordinate(geo, A.vhit_hit->WireID());
+  double B_w = T.WireCoordinate(geo, B.vhit_hit->WireID());
+
+  // Use this to get y, z
+  double y = T.TwoPlaneToY(geo->View(A.vhit_hit->WireID()), A_w, geo->View(B.vhit_hit->WireID()), B_w);
+  double z = T.TwoPlaneToZ(geo->View(A.vhit_hit->WireID()), A_w, geo->View(B.vhit_hit->WireID()), B_w);
+
+  // just average the x-pos
+  double x = (dprop.ConvertTicksToX(A.vhit_hit->PeakTime(), A.vhit_hit->WireID()) + dprop.ConvertTicksToX(B.vhit_hit->PeakTime(), B.vhit_hit->WireID())) / 2.;
+
+  geo::Point_t pos(x, y, z);
```
I don't really follow what you mean here.

PetrilloAtWork replied:
My comment is in two parts: about _x_ and about _y_ × _z_.

For _x_: let's make an example... hit times are about `0`, both of them. `dprop.ConvertTicksToX(0, plane)` returns `0`, doesn't it? So averaging the two you get `x = 0`. What is _x_ = 0 supposed to mean?

For _y_ × _z_: saying that there is a LArSoft alternative to the `TwoPlaneToY()`/`TwoPlaneToZ()` functions, provided that I understand correctly that the purpose of that code is to find the geometric "intersection" of two wires.

gputnam replied:
Ok! I understand now about "x". I believe this is another assumption I have made that works in SBND but not ICARUS. I will add it to the list.

On y,z -- is there such a function? I could not find any.

PetrilloAtWork replied:
That would be [`geo::GeometryCore::WireIDsIntersect()`](https://nusoft.fnal.gov/larsoft/doxsvn/html/classgeo_1_1GeometryCore.html#a6e7e4d4d21650dfa9d9f84260465ba93).
Maybe you can take a look at its documentation and decide.

PetrilloAtWork replied:
I have understood my mistake on the first half of the comment: `dprop.ConvertTicksToX(0, plane)` does not return `0`. I thought it would return a _x_ coordinate _relative_ to the wire plane, but instead it returns an absolute coordinate. I have fixed the suggestion above to reflect that.
Given that, I think that _x_ is ok both for ICARUS and SBND.

gputnam commented on sbncode/TPCReco/VertexStub/TwoPlaneStubMerge_tool.cc:
```diff
@@ -0,0 +1,220 @@
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+TwoPlaneStubMerge::~TwoPlaneStubMerge()
+{
+}
```
I think it's fine as a placeholder -- is this any different than an implicit destructor?

PetrilloAtWork replied:
... maybe.
First, you don't declare it `noexcept`, which the default destructor would be. Compiler might be smart enough to figure out though.
Then, there are a lot of rules of C++ about which constructors and assignment operators are implicitly declared when some others are present (e.g. if a user-defined move constructor is present, a copy assignment operator is not generated; and vice versa; if a user-defined destructor is present, the move constructor is not implicitly declared). I am not sure what is the discretion of the compiler here; in general, I prefer to just avoid the mess completely and let the compiler the most free.

Additionally, if you add _content_ in the destructor and I am reviewing the changes, I will then try to persuade you to rewrite the class so that the explicit destruction is not needed. That is possible most of the times in C++.

gputnam replied:
Fixed.

gputnam commented on sbncode/TPCReco/VertexStub/TwoPlaneStubMerge_tool.cc:
```diff
@@ -0,0 +1,220 @@
+  }
+
+  // Save the stubs which were not merged across planes and not marked to be removed
+  for (unsigned i_stub = 0; i_stub < stubs.size(); i_stub++) {
+    if (fSaveOldStubs || (!hasmerged.count(i_stub) && !dontsave.count(i_stub))) {
+      ret.push_back(stubs[i_stub]);
+    }
```
Yes -- that's the intended behavior.

gputnam commented on sbncode/TPCReco/VertexStub/VertexStubTracker_module.cc:
```diff
@@ -66,19 +69,29 @@ class sbn::VertexStubTracker : public art::EDProducer {
 {
+  // load the tools
+  std::vector<fhicl::ParameterSet> merge_tool_configs(p.get<std::vector<fhicl::ParameterSet>>("MergeTools"));
+  for (unsigned i = 0; i < merge_tool_configs.size(); i++) {
+    fStubMergeTools.push_back(art::make_tool<IStubMerge>(merge_tool_configs[i]));
```
see previous.

gputnam commented on sbncode/TPCReco/VertexStub/VertexStubTracker_module.cc:
```diff
@@ -181,95 +120,71 @@ void sbn::VertexStubTracker::produce(art::Event& e)
-
-  std::vector<art::Ptr<sbn::VertexHit>> vhits;
-  art::fill_ptr_vector(vhits, vhit_handle);
+  art::Handle<std::vector<recob::Slice>> slice_handle;
+  e.getByLabel(fPFPLabel, slice_handle);
```
As said earlier I prefer the existing method.

gputnam commented:
Ok -- I have done a first pass of going through the comments. I disagreed with some of the style-related comments. There are a number where I made a change, I will commit this presently. 

I think the unresolved issues are on the plane matching and on documentation. I will go through the code in the next couple days to add more documentation. I am open to more discussion on how exactly the plane matching should work.

gputnam commented on sbncode/TPCReco/NuVertexChargeTree_module.cc:
```diff
@@ -642,17 +852,28 @@ void sbn::NuVertexChargeTree::FillVertexHits(
    const std::vector<std::vector<std::pair<int, float>>> &slice_pfp_matches,
    const std::vector<art::Ptr<simb::MCParticle>> &g4_mcparticles, 
-   const std::vector<const sim::GeneratedParticleInfo *> infos) {
+   const std::vector<const sim::GeneratedParticleInfo *> infos,
+   const geo::GeometryCore *geo,
```
Fixed.

gputnam commented on sbncode/TPCReco/NuVertexChargeTree_module.cc:
```diff
@@ -706,11 +936,74 @@ void sbn::NuVertexChargeTree::FillNeutrino(const simb::MCTruth &nu,
+    float thisQ = 0.;
+
+    std::map<geo::WireID, float> chargemap;
+    std::set<geo::View_t> views_set = geo->Views();
+    std::vector<geo::View_t> views;
+    for (auto const &v: views_set) views.push_back(v);
+    for (geo::View_t view: views) {
```
Fixed.

gputnam commented on sbncode/TPCReco/NuVertexChargeTree_module.cc:
```diff
@@ -706,11 +936,74 @@ void sbn::NuVertexChargeTree::FillNeutrino(const simb::MCTruth &nu,
+
+        // Correct for electron lifetime 
+        geo::Point_t p {ide->x, ide->y, ide->z};
+        geo::TPCID tpc = geo->FindTPCAtPosition(p);
+        if (tpc) {
+          float driftT = abs(ide->x - geo->TPC(tpc).PlaneLocation(0)[0]) / dprop.DriftVelocity();
```
Fixed.

gputnam commented on sbncode/TPCReco/NuVertexChargeTree_module.cc:
```diff
@@ -706,11 +936,74 @@ void sbn::NuVertexChargeTree::FillNeutrino(const simb::MCTruth &nu,
+          float driftT = abs(ide->x - geo->TPC(tpc).PlaneLocation(0)[0]) / dprop.DriftVelocity();
+          thisQ += ide->numElectrons * exp(driftT / dprop.ElectronLifetime());
+
+          int plane = -1;
+          for (unsigned i_plane = 0; i_plane < 3; i_plane++) {
+            geo::PlaneID planeID {tpc, i_plane};
+            if (geo->View(planeID) == view) {
+              plane = i_plane;
+              break;
+            }
+          }
+
+          geo::PlaneID planeID {tpc, (unsigned)plane};
+          try {
+            geo::WireID::WireID_t wid = geo->NearestWire(p, planeID);
+            geo::WireID w {planeID, wid};
+            chargemap[w] += ide->numElectrons * exp(driftT / dprop.ElectronLifetime());
+          }
+          catch(...) {}
```
Fixed.

gputnam commented on sbncode/TPCReco/NuVertexChargeTree_module.cc:
```diff
@@ -980,6 +1275,18 @@ void sbn::NuVertexChargeTree::analyze(art::Event const& evt)
+    for (unsigned i = 0; i < thisStubs.size(); i++) {
+      thisStubInfos.emplace_back();
+
+      thisStubInfos.back().stub = *thisStubs[i];
+      thisStubInfos.back().pfp = thisStubPFPs.at(i);
+      thisStubInfos.back().hits = thisStubHits.at(i);
+      thisStubInfos.back().vhit = thisVertexHits.at(thisStubHitInds[i]);
+      thisStubInfos.back().vhit_hit = thisVertexHitHits.at(thisStubHitInds[i]); 
```
Fixed.

gputnam commented on sbncode/TPCReco/VertexStub/CMakeLists.txt:
```diff
@@ -1,3 +1,30 @@
+	art_Persistency_Provenance
+	art_Utilities canvas
+	${MF_MESSAGELOGGER}
+	${MF_UTILITIES}
```
Fixed.

gputnam replied:
Some of them were needed though.

PetrilloAtWork replied:
I am afraid I might have messed the selection when writing the comments... glad you could sort it out.
I think `${MF_UTILITIES}` translates into nothing, but the others do at very least exist.
Resolved on my side.

gputnam commented on sbncode/TPCReco/VertexStub/CMakeLists.txt:
```diff
@@ -1,3 +1,30 @@
+	${FHICLCPP}
+	cetlib cetlib_except
+	${ROOT_EVE_LIB_LIST}
+	${ROOT_X3d}
```
Fixed.

gputnam commented on sbncode/TPCReco/VertexStub/CMakeLists.txt:
```diff
@@ -1,3 +1,30 @@
+        LIBRARIES
+	${ART_FRAMEWORK_CORE}
+	${ART_FRAMEWORK_SERVICES_REGISTRY}
+	${ART_ROOT_IO_TFILESERVICE_SERVICE}                ${ART_FRAMEWORK_SERVICES}
```
Some of these were needed, fixed otherwise.

gputnam commented on sbncode/TPCReco/VertexStub/CMakeLists.txt:
```diff
@@ -1,3 +1,30 @@
+	sbnobj_Common_Reco
+	lardataobj_AnalysisBase
+        larcorealg_Geometry
+        larcore_Geometry_Geometry_service
```
Some of these I believe are needed.

PetrilloAtWork replied:
The comment was meant to cover only `larcore_Geometry_Geometry_service`.
I may need some GitHub training...

gputnam commented on sbncode/TPCReco/VertexStub/IStubMerge.h:
```diff
@@ -0,0 +1,55 @@
+
+// cpp includes
+#include <vector>
+#include <array>
```
Fixed.

gputnam commented on sbncode/TPCReco/VertexStub/PlaneStubMerge_tool.cc:
```diff
@@ -0,0 +1,97 @@
+
+// std includes
+#include <string>
+#include <iostream>
```
Fixed.

gputnam commented on sbncode/TPCReco/VertexStub/PlaneTransform.h:
```diff
@@ -0,0 +1,89 @@
+#include "sbnobj/Common/Reco/VertexHit.h"
+#include "sbnobj/Common/Reco/Stub.h"
+
+#include <memory>
```
Fixed.

gputnam commented on sbncode/TPCReco/VertexStub/PlaneStubMerge_tool.cc:
```diff
@@ -0,0 +1,97 @@
+#include "art/Framework/Services/Registry/ServiceHandle.h"
+#include "art/Persistency/Common/PtrMaker.h"
+#include "art/Utilities/ToolMacros.h"
+#include "cetlib/cpu_timer.h"
```
Fixed.

gputnam commented on sbncode/TPCReco/VertexStub/PlaneStubMerge_tool.cc:
```diff
@@ -0,0 +1,97 @@
+#include "larcorealg/Geometry/BoxBoundedGeo.h"
+#include "larcore/Geometry/Geometry.h"
+#include "larcorealg/Geometry/GeometryCore.h"
+#include "larcore/CoreUtils/ServiceUtil.h"
```
Fixed.

gputnam commented on sbncode/TPCReco/VertexStub/StubBuilder.cxx:
```diff
@@ -0,0 +1,232 @@
+    fSliceTrkTHMs[slices[i_slc].key()] = thisSliceTrkTHMs;
+    fSliceTrks[slices[i_slc].key()] = thisSliceTracks;
+    fSlicePFPs[slices[i_slc].key()] = thisSlicePFPs;
+    fSliceHits[slices[i_slc].key()] = sliceHits.at(i_slc);
```
Fixed.

gputnam commented on sbncode/TPCReco/VertexStub/StubBuilder.h:
```diff
@@ -0,0 +1,63 @@
+#include "lardataobj/RecoBase/Track.h"
+#include "lardataobj/RecoBase/TrackTrajectory.h"
+#include "lardataobj/RecoBase/Trajectory.h"
+#include "lardataobj/RecoBase/Wire.h"
```
Fixed.

gputnam commented on sbncode/TPCReco/VertexStub/StubBuilder.h:
```diff
@@ -0,0 +1,63 @@
+#include "sbncode/TPCReco/VertexStub/StubMergeAlgorithms.h"
+
+#include <memory>
+
```
Fixed.

gputnam commented on sbncode/TPCReco/VertexStub/StubMergeAlgorithms.cxx:
```diff
@@ -0,0 +1,245 @@
+    // Gets relative E field Distortions
+    geo::Vector_t EFieldOffsets = sce->GetEfieldOffsets(loc);
+    // Add 1 in X direction as this is the direction of the drift field
+    EFieldOffsets = EFieldOffsets + geo::Vector_t{1, 0, 0};
```
Fixed.

gputnam commented on sbncode/TPCReco/VertexStub/StubMergeAlgorithms.cxx:
```diff
@@ -0,0 +1,245 @@
+    geo::View_t view, geo::TPCID tpc, 
+    bool correct_sce, bool track_is_sce_corrected, float xsign) {
+
+  double angleToVert = geo->WireAngleToVertical(view, tpc.TPC, tpc.Cryostat) - 0.5*::util::pi<>();
```
Fixed.

gputnam commented on sbncode/TPCReco/VertexStub/StubMergeAlgorithms.cxx:
```diff
@@ -0,0 +1,245 @@
+  const TVector3 &dir_corr {pitch*dir_w.X() + (dirOffsets.X() - locOffsets.X()),  
+                            pitch*dir_w.Y() + (dirOffsets.Y() - locOffsets.Y()), pitch*dir_w.Z() + (dirOffsets.Z() - locOffsets.Z())};
+
+  pitch = dir_corr.Mag();
```
Fixed.

gputnam commented on sbncode/TPCReco/VertexStub/StubMergeAlgorithms.cxx:
```diff
@@ -0,0 +1,245 @@
+  geo::Vector_t dirOffsets = {0., 0., 0.};
+  geo::Vector_t locOffsets = {0., 0., 0.};
+  if (sce && sce->EnableCalSpatialSCE() && correct_sce) {
+    dirOffsets = sce->GetCalPosOffsets(geo::Point_t{loc_w.X()  + pitch*dir_w.X(), loc_w.Y() + pitch*dir_w.Y(), loc_w.Z() + pitch*dir_w.Z()}, tpc.TPC);
```
Fixed.

gputnam commented on sbncode/TPCReco/VertexStub/StubMergeAlgorithms.cxx:
```diff
@@ -0,0 +1,245 @@
+    loc_pdx.SetY(loc_pdx.Y() + loc_pdx_offset.Y());
+    loc_pdx.SetZ(loc_pdx.Z() + loc_pdx_offset.Z());
+
+    dir_w = (loc_pdx - loc_mdx) /  (loc_mdx - loc_pdx).r(); 
```
Fixed.

gputnam commented on sbncode/TPCReco/VertexStub/StubMergeAlgorithms.cxx:
```diff
@@ -0,0 +1,245 @@
+    const spacecharge::SpaceCharge *sce,
+    const detinfo::DetectorPropertiesData &dprop) {
+
+  static const bool VERBOSE = false;
```
Fixed.

gputnam commented on sbncode/TPCReco/VertexStub/StubMergeAlgorithms.h:
```diff
@@ -0,0 +1,69 @@
+#include "larcorealg/Geometry/GeometryCore.h"
+#include "lardataalg/DetectorInfo/DetectorPropertiesStandard.h"
+#include "lardata/DetectorInfoServices/DetectorPropertiesService.h"
+#include "lardataalg/DetectorInfo/DetectorClocksStandard.h"
```
Fixed.

gputnam commented on sbncode/TPCReco/VertexStub/StubMergeAlgorithms.h:
```diff
@@ -0,0 +1,69 @@
+#ifndef StubMergeAlgorithms_HH
+#define StubMergeAlgorithms_HH
+
+#include "larcore/Geometry/Geometry.h"
```
Fixed.

gputnam commented on sbncode/TPCReco/VertexStub/StubMergeAlgorithms.h:
```diff
@@ -0,0 +1,69 @@
+#include "lardataalg/DetectorInfo/DetectorClocksStandard.h"
+
+#include "larevt/SpaceCharge/SpaceCharge.h"
+#include "larevt/SpaceChargeServices/SpaceChargeService.h"
```
Fixed.

gputnam commented on sbncode/TPCReco/VertexStub/StubMergeAlgorithms.h:
```diff
@@ -0,0 +1,69 @@
+
+#include "sbncode/TPCReco/VertexStub/PlaneTransform.h"
+
+#include <memory>
```
Fixed. PlaneTransform is needed. Removed memory.

gputnam commented on sbncode/TPCReco/VertexStub/TwoPlaneStubMerge_tool.cc:
```diff
@@ -0,0 +1,220 @@
+      if (A.stub.plane.size() != 1 || B.stub.plane.size() != 1) continue; // Only match 1-plane-stubs
+
+      if (A.stub.plane.front().Plane == B.stub.plane.front().Plane) continue; // Match across planes
+      if (A.stub.plane.front().TPC != B.stub.plane.front().TPC) continue; // But should be in the same TPC
```
Fixed.

gputnam commented on sbncode/TPCReco/VertexStub/TwoPlaneStubMerge_tool.cc:
```diff
@@ -0,0 +1,220 @@
+  // Output list
+  std::vector<sbn::StubInfo> ret;
+
+  for (unsigned i_mrg = 0; i_mrg < merges.size(); i_mrg++) {
+    const MergeInfo &thismrg = merges[i_mrg];
```
Again I think the existing for loop is clear and fine.

gputnam commented on sbncode/TPCReco/VertexStub/VertexStubTracker_module.cc:
```diff
@@ -87,89 +100,15 @@ sbn::VertexStubTracker::VertexStubTracker(fhicl::ParameterSet const& p)
   auto const dprop =
     art::ServiceHandle<detinfo::DetectorPropertiesService const>()->DataFor(e, clock_data);
+  // TODO: fix -- for now, use a null space-charge service
+  const spacecharge::SpaceCharge *sce = NULL;
```
Fixed.

gputnam commented on sbncode/TPCReco/VertexStub/VertexStubTracker_module.cc:
```diff
@@ -181,95 +120,71 @@ void sbn::VertexStubTracker::produce(art::Event& e)
-    // Collect more data
-    art::Ptr<recob::Slice> thisSlice = vhitSlcs.at(i_vhit).at(0);
+  for (unsigned i_slc = 0; i_slc < slices.size(); i_slc++) {
+    const std::vector<art::Ptr<sbn::VertexHit>> vhits = slcVHits.at(i_slc);
```
Fixed.

gputnam commented on sbncode/TPCReco/VertexStub/VertexStubTracker_module.cc:
```diff
@@ -181,95 +120,71 @@ void sbn::VertexStubTracker::produce(art::Event& e)
+    art::Ptr<recob::Slice> thisSlice = slices[i_slc];
+    // each hit should have the same vertex
+    recob::Vertex vertex;
+    if (vhits.size()) vertex = *vhitVtxs.at(0).at(0);
```
I think it's fine/expressive as is.

PetrilloAtWork replied:
I challenge that: `if (vhits.size())` is not as expressive as `if (!vhits.empty())`. `empty()` exactly expresses what you are interested in; `size()` does not.

gputnam replied:
Fixed.

gputnam commented:
Done! I have addressed the first round of comments (except for more documentation, which I will do soon).

gputnam commented:
Added overview documentation.

PetrilloAtWork reviewed (changes requested):
Gone through the replies, I hope I did not miss any.
Some conversations are still open.

PetrilloAtWork commented:
My last round of comment did not include the last commits about the documentation.
Good job, but would it possible to make it Doxygen'ed? I can help if needed.

gputnam commented on sbncode/TPCReco/NuVertexChargeTree_module.cc:
```diff
@@ -706,11 +936,74 @@ void sbn::NuVertexChargeTree::FillNeutrino(const simb::MCTruth &nu,
-    for (const sim::IDE *ide: particle_ides) thisVisE += ide->energy;
+    float thisQ = 0.;
+
+    std::map<geo::WireID, float> chargemap;
```
It may simplify the way chargemap is currently used. But I also build that map to enable other possible development of truth based variables. So changing the way chargemap works would make future development more difficult, without an appreciable performance gain.

gputnam commented on sbncode/TPCReco/VertexStub/TwoPlaneStubMerge_tool.cc:
```diff
@@ -0,0 +1,220 @@
+/**
+ *
```
Added more credit.

gputnam commented:
I am back online! And just did another round of addressing comments. Would you be able to say what is left to address? It is a little hard for me to follow everything.

And re-doxygen -- that would be great! Would definitely appreciate the help.

gputnam commented:
And here is the list of things to fix for ICARUS:

-correctly account for space charge -- don't make assumption about x==0 location
-ConvertTicktoX doesn't give correct x position in detector geometry
-don't make assumptions about plane numbering
-Get correct wire orientation config

gputnam commented on sbncode/TPCReco/VertexStub/PlaneStubMerge_tool.cc:
```diff
@@ -0,0 +1,97 @@
+    /**
+     *  @brief  Destructor
+     */
+    ~PlaneStubMerge();
```
Fixed.

gputnam commented on sbncode/TPCReco/VertexStub/PlaneStubMerge_tool.cc:
```diff
@@ -0,0 +1,97 @@
+PlaneStubMerge::~PlaneStubMerge()
+{
+}
+
```
Fixed.

gputnam commented on sbncode/TPCReco/NuVertexChargeTree_module.cc:
```diff
@@ -642,17 +852,28 @@ void sbn::NuVertexChargeTree::FillVertexHits(
 
   art::ServiceHandle<cheat::BackTrackerService> backtracker;
+  // TODO: fix -- for now, use a null space-charge service
+  const spacecharge::SpaceCharge *sce = NULL;
```
Fixed.

PetrilloAtWork reviewed (approved):
As far as I can tell, all the essentials have been addressed (together with _a lot_ more).
Thank you for your patience and your work!

FNALbuild commented:
CI build for LArSoft on slf7 for c7:prof is in progress -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app//view_builds/index?offset=0&builds=lar_ci/11706&builds=)

FNALbuild commented:
CI build for LArSoft on slf7 for e20:prof is in progress -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app//view_builds/index?offset=0&builds=lar_ci/11705&builds=)

FNALbuild commented:
:heavy_check_mark: CI build for LArSoft Succeeded on slf7 for e20:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app//view_builds/index?offset=0&builds=lar_ci/11705&builds=)

FNALbuild commented:
CI build for ICARUS on slf7 for e20:prof is in progress -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/view_builds/index?offset=0&builds=icarus_ci/2059&builds=)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/11705&builds=)

FNALbuild commented:
CI build for SBND on slf7 for e20:prof is in progress -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/view_builds/index?offset=0&builds=sbnd_ci/6589&builds=)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/11705&builds=)

FNALbuild commented:
:x: CI build for ICARUS Failed at phase ci_tests ICARUS on slf7 for e20:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/view_builds/index?offset=0&builds=icarus_ci/2059&builds=)

:rotating_light: For more details about the failed phase, check the [ci_tests ICARUS phase logs](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/build_detail/phase_details?build_id=icarus_ci/2059&platform=Linux%203.10.0-1160.31.1.el7.x86_64&phase=ci_tests&buildtype=slf7%20e20:prof)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/11705&builds=)

FNALbuild commented:
:warning: CI build for SBND Warning at phase ci_tests SBND on slf7 for e20:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/view_builds/index?offset=0&builds=sbnd_ci/6589&builds=)

:rotating_light: For more details about the warning phase, check the [ci_tests SBND phase logs](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/build_detail/phase_details?build_id=sbnd_ci/6589&platform=Linux%203.10.0-1160.31.1.el7.x86_64&phase=ci_tests&buildtype=slf7%20e20:prof)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/11705&builds=)

FNALbuild commented:
:heavy_check_mark: CI build for LArSoft Succeeded on slf7 for c7:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app//view_builds/index?offset=0&builds=lar_ci/11706&builds=)

FNALbuild commented:
CI build for ICARUS on slf7 for c7:prof is in progress -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/view_builds/index?offset=0&builds=icarus_ci/2061&builds=)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/11706&builds=)

FNALbuild commented:
CI build for SBND on slf7 for c7:prof is in progress -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/view_builds/index?offset=0&builds=sbnd_ci/6591&builds=)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/11706&builds=)

FNALbuild commented:
:x: CI build for ICARUS Failed at phase build ICARUS on slf7 for c7:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/view_builds/index?offset=0&builds=icarus_ci/2061&builds=)

:rotating_light: For more details about the failed phase, check the [build ICARUS phase logs](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/build_detail/phase_details?build_id=icarus_ci/2061&platform=Linux%203.10.0-1160.31.1.el7.x86_64&phase=build&buildtype=slf7%20c7:prof)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/11706&builds=)

FNALbuild commented:
:x: CI build for SBND Failed at phase build SBND on slf7 for c7:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/view_builds/index?offset=0&builds=sbnd_ci/6591&builds=)

:rotating_light: For more details about the failed phase, check the [build SBND phase logs](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/build_detail/phase_details?build_id=sbnd_ci/6591&platform=Linux%203.10.0-1160.31.1.el7.x86_64&phase=build&buildtype=slf7%20c7:prof)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/11706&builds=)

FNALbuild commented:
CI build for LArSoft on slf7 for e20:prof is in progress -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app//view_builds/index?offset=0&builds=lar_ci/11718&builds=)

FNALbuild commented:
CI build for LArSoft on slf7 for c7:prof is in progress -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app//view_builds/index?offset=0&builds=lar_ci/11719&builds=)

FNALbuild commented:
:heavy_check_mark: CI build for LArSoft Succeeded on slf7 for c7:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app//view_builds/index?offset=0&builds=lar_ci/11719&builds=)

FNALbuild commented:
:heavy_check_mark: CI build for LArSoft Succeeded on slf7 for e20:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app//view_builds/index?offset=0&builds=lar_ci/11718&builds=)

FNALbuild commented:
CI build for SBND on slf7 for c7:prof is in progress -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/view_builds/index?offset=0&builds=sbnd_ci/6599&builds=)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/11719&builds=)

FNALbuild commented:
CI build for SBND on slf7 for e20:prof is in progress -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/view_builds/index?offset=0&builds=sbnd_ci/6600&builds=)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/11718&builds=)

FNALbuild commented:
CI build for ICARUS on slf7 for e20:prof is in progress -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/view_builds/index?offset=0&builds=icarus_ci/2064&builds=)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/11718&builds=)

FNALbuild commented:
CI build for ICARUS on slf7 for c7:prof is in progress -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/view_builds/index?offset=0&builds=icarus_ci/2065&builds=)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/11719&builds=)

FNALbuild commented:
:x: CI build for ICARUS Failed at phase build ICARUS on slf7 for c7:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/view_builds/index?offset=0&builds=icarus_ci/2065&builds=)

:rotating_light: For more details about the failed phase, check the [build ICARUS phase logs](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/build_detail/phase_details?build_id=icarus_ci/2065&platform=Linux%203.10.0-1160.31.1.el7.x86_64&phase=build&buildtype=slf7%20c7:prof)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/11719&builds=)

FNALbuild commented:
:x: CI build for ICARUS Failed at phase ci_tests ICARUS on slf7 for e20:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/view_builds/index?offset=0&builds=icarus_ci/2064&builds=)

:rotating_light: For more details about the failed phase, check the [ci_tests ICARUS phase logs](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/build_detail/phase_details?build_id=icarus_ci/2064&platform=Linux%203.10.0-1160.31.1.el7.x86_64&phase=ci_tests&buildtype=slf7%20e20:prof)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/11718&builds=)

FNALbuild commented:
:warning: CI build for SBND Warning at phase ci_tests SBND on slf7 for c7:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/view_builds/index?offset=0&builds=sbnd_ci/6599&builds=)

:rotating_light: For more details about the warning phase, check the [ci_tests SBND phase logs](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/build_detail/phase_details?build_id=sbnd_ci/6599&platform=Linux%203.10.0-1160.31.1.el7.x86_64&phase=ci_tests&buildtype=slf7%20c7:prof)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/11719&builds=)

FNALbuild commented:
:heavy_check_mark: CI build for SBND Succeeded on slf7 for e20:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/view_builds/index?offset=0&builds=sbnd_ci/6600&builds=)


parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/11718&builds=)
